pallets/nfts/src/features/approvals.rs

```diff
@@ -165,10 +181,191 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+			collection_config.is_setting_enabled(CollectionSetting::TransferableItems),
+			Error::<T, I>::ItemsNonTransferable
+		);
+		let now = frame_system::Pallet::<T>::block_number();
+		let deadline = maybe_deadline.map(|d| d.saturating_add(now));
+
+		CollectionApprovals::<T, I>::try_mutate_exists(
+			(&collection, &owner, &delegate),
+			|maybe_approval| -> DispatchResult {
+				let deposit_required = T::CollectionApprovalDeposit::get();
+				let current_deposit =
+					maybe_approval.take().map(|(_, deposit)| deposit).unwrap_or_default();
+				if current_deposit < deposit_required {
+					T::Currency::reserve(&owner, deposit_required - current_deposit)?;
+				}
+				*maybe_approval = Some((deadline, deposit_required));
+				Ok(())
+			},
+		)?;
+
+		Self::deposit_event(Event::TransferApproved {
+			collection,
+			item: None,
+			owner,
+			delegate,
+			deadline,
+		});
+
+		Ok(())
+	}
+
+	/// Cancels the transfer of all `collection` items of `owner` to a `delegate`.
+	///
+	/// This function is used to cancel the approval for the transfer of the collection items of
+	/// `owner` to a `delegate`. After canceling the approval, the deposit is returned to the
+	/// `owner` account and the `ApprovalCancelled` event is emitted.
+	///
+	/// - `owner`: The owner of the collection items.
+	/// - `collection`: The identifier of the collection.
+	/// - `delegate`: The account that had permission to transfer collection items.
+	pub(crate) fn do_cancel_collection_approval(
+		owner: T::AccountId,
+		collection: T::CollectionId,
+		delegate: T::AccountId,
+	) -> DispatchResult {
+		let (_, deposit) = CollectionApprovals::<T, I>::take((&collection, &owner, &delegate))
+			.ok_or(Error::<T, I>::Unapproved)?;
+
+		T::Currency::unreserve(&owner, deposit);
+
+		Self::deposit_event(Event::ApprovalCancelled { collection, owner, item: None, delegate });
+
+		Ok(())
+	}
+
+	/// Clears all collection approvals.
+	///
+	/// This function is used to clear `limit` collection approvals for the
+	/// collection items of `owner`. After clearing all approvals, the deposit of each collection
+	/// approval is returned to the `owner` account and the `ApprovalsCancelled` event is
+	/// emitted.
+	///
+	/// - `owner`: The owner of the collection items.
+	/// - `collection`: The identifier of the collection.
+	/// - `limit`: The amount of collection approvals that will be cleared.
+	pub(crate) fn do_clear_collection_approvals(
+		owner: T::AccountId,
+		collection: T::CollectionId,
+		limit: u32,
+	) -> Result<u32, DispatchError> {
+		if limit == 0 {
+			return Ok(0);
+		}
+		let mut removed_approvals: u32 = 0;
+		let mut deposits: BalanceOf<T, I> = Zero::zero();
+		// Iterate and remove each collection approval, returning the deposit back to the `owner`.
+		for (_, (_, deposit)) in CollectionApprovals::<T, I>::drain_prefix((collection, &owner)) {
+			deposits = deposits.saturating_add(deposit);
+			removed_approvals.saturating_inc();
+			if removed_approvals == limit {
```

Comment:
Can we not eliminate this with `drain_prefix((collection, &owner)).take(limit)`?

Can probably also use `deposits.saturating_accrue(deposit)`

Reply:
Thank you, great suggestion! Resolved.